rfcs/text/0012_background_sessions.md

```diff
@@ -0,0 +1,291 @@
+
+Because a single view (such as a dashboard with multiple visualizations) can initiate multiple search requests, we need
+a way to associate the search requests together in a single entity. We call this entity a `session`, and when a user
+decides that they want to continue running the search requests while moving onto other work, we will create a saved
+object corresponding with that specific `session`.
+
+We will expose a new `session` service inside of the `data` plugin `search` service. The service will expose APIs for 
+the following:
+
+- Starting a new session (generating a unique identifier for this specific grouping of search requests)
+- Saving a session (creating the saved object using that ID)
+- Getting a list of sessions for the current user (including status, any necessary application data, and a URL to
+restore the view)
+- Cancelling an incomplete session (deleting the saved object and cancelling any incomplete search requests)
+- Restoring a session (allowing search requests to resume instead of restart)
+
+We will also modify the existing `search` API exposed by the `data` plugin `search` service to optionally accept an
+additional parameter, the `sessionId`, which will allow search requests to *resume* instead of *restart*. When
+"restoring" a session, it is crucial that search requests are sent with *identical* search request parameters as were
+sent originally. (For example, relative date ranges should be converted to absolute before passing the request to the
+`search` service.)
+
+Applications query data from Elasticsearch on behalf of the end user in a variety of ways:
+
+- Using `SearchSource` (our higher-level abstracted search API, used by Maps, Lens, and most visualizations, exposed by
+the `data` plugin `search` service)
+- Using `search` (our lower-level search API that accepts raw Elasticsearch query DSL, used by Vega and Timelion,
+exposed by the `data` plugin `search` service)
+- Exposing a route that bypasses our `search` services and queries Elasticsearch directly, then manipulates the data
+before returning a response (as in the case of TSVB and most solutions)
+
+The proposed APIs provide the tools necessary to run long search requests, view their status, cancel incomplete search
+requests, and retrieve results, regardless of the mechanism used to query Elasticsearch.
+
+## Background session service
+
+The primary service which will facilitate background sessions is the server-side `session` service.
+
+When a `search` request is handled, and it contains a `sessionId`, the service will query Elasticsearch to see if there
+is already a saved object for that session. If there isn't, then we will forward the search request to Elasticsearch and
+receive an ID specific to that search request. We store that ID in memory, keyed off of a hash of the request parameters
+and the current username.
+
+We will then periodically query Elasticsearch to see if there is a saved object for that session (meaning the user has
```

Comment:
If that Kibana process is restarted during this period, the search ID will be lost from memory, and it won't be added to the saved-object. Is my understanding correct that this will prevent the background search from being restored and consumed entirely? Wouldn't partial use of the stored background search be more user-friendly?

Reply:
Actually I think you would still be able to restore the background search, but those that were lost would result in errors. (In other words, some panels could still load successfully.)